mqtt/mqtt-edgehub/src/auth/authorization/edgehub.rs

```diff
@@ -0,0 +1,338 @@
+    ClientId, ClientInfo,
+};
+
+const FORBIDDEN_TOPIC_FILTERS: [&str; 3] = ["#", "$upstream/", "$downstream/"];
+
+const FORBIDDEN_TOPIC_PREFIXES: [&str; 3] = ["$SYS/", "$upstream/", "$downstream/"];
+
+#[derive(Debug, Default)]
+pub struct EdgeHubAuthorizer;
+
+#[allow(clippy::unused_self)]
+impl EdgeHubAuthorizer {
+    fn authorize_connect(
+        &self,
+        client_id: &ClientId,
+        client_info: &ClientInfo,
+        _connect: &Connect,
+    ) -> Authorization {
+        match client_info.auth_id() {
+            // forbid anonymous clients to connect to the broker
+            AuthId::Anonymous => {
+                Authorization::Forbidden("Anonymous clients cannot connect to broker".to_string())
+            }
+            // allow only those clients whose auth_id and client_id identical
+            AuthId::Identity(identity) => {
+                if identity == client_id.as_str() {
+                    Authorization::Allowed
+                } else {
+                    Authorization::Forbidden(format!(
+                        "client_id {} does not match registered iothub identity id.",
+                        client_id
+                    ))
+                }
+            }
+        }
+    }
+
+    fn authorize_publish(
+        &self,
+        client_id: &ClientId,
+        client_info: &ClientInfo,
+        publish: &Publish,
+    ) -> Authorization {
+        let topic = &publish.publication.topic_name;
+
+        if is_forbidden_topic(topic) {
+            Authorization::Forbidden(format!("{} is forbidden topic filter", topic))
+        } else if is_iothub_topic(topic) {
+            // run authorization rules for publication to IoTHub topic
+            self.authorize_iothub_topic(client_id, client_info, topic)
+        } else {
+            // allow any client to publish to any non-iothub topics
+            Authorization::Allowed
+        }
+    }
+
+    fn authorize_subscribe(
+        &self,
+        client_id: &ClientId,
+        client_info: &ClientInfo,
+        subscribe: &Subscribe,
+    ) -> Authorization {
+        let topic = subscribe.topic_filter();
+
+        if is_forbidden_topic_filter(topic) {
+            Authorization::Forbidden(format!("{} is forbidden topic filter", topic))
+        } else if is_iothub_topic(topic) {
+            // run authorization rules for subscription to IoTHub topic
+            self.authorize_iothub_topic(client_id, client_info, topic)
+        } else {
+            // allow any client to subscribe to any non-iothub topics
+            Authorization::Allowed
+        }
+    }
+
+    fn authorize_iothub_topic(
+        &self,
+        client_id: &ClientId,
+        client_info: &ClientInfo,
+        topic: &str,
+    ) -> Authorization {
+        match client_info.auth_id() {
+            // forbid anonymous clients to subscribe to restricted topics
+            AuthId::Anonymous => Authorization::Forbidden(
```

Comment:
Do we need this explicitly (since we don't let anon clients to connect)?

Reply:
Correct, but I'd like to check in place. We may want to let anon clients connect to access generic MQTT topics but forbid iothub related topics.